### PR TITLE
Implement __cxa_thread_atexit to support C++11 thread_local.

### PR DESCRIPTION
**Pull request by Flast**
This is tiny implementation because it doesn't postpone unload .so (i.e. `dlclose`) like a glibc's [`__cxa_thread_atexit_impl`](https://sourceware.org/glibc/wiki/Destructor%20support%20for%20thread_local%20variables).

Tested: FreeBSD10.2 x86_64 (clang 3.4.1), Fedora 24 rawhide x86_64 (gcc 5.3.1).

### CODE REVIEW

File: src/exception.cc [hunk 1]
```diff
@@ -318,110 +268,11 @@ static void dependent_exception_cleanup(_Unwind_Reason_Code reason,
...
-	pthread_key_create(&eh_key, thread_cleanup);
-	pthread_setspecific(eh_key, reinterpret_cast<void *>(0x42));
-	fakeTLS = (pthread_getspecific(eh_key) != reinterpret_cast<void *>(0x42));
-	pthread_setspecific(eh_key, 0);
-}
-
-/**
- * Returns the thread info structure, creating it if it is not already created.
- */
-static __cxa_thread_info *thread_info()
```
Comment (davidchisnall) [inline src/exception.cc]: These functions (at least, the fast path) are expected to be amenable for inlining.  Moving them defeats this.

File: src/thread_info.cc [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+{
+	if (fakeTLS) { return &singleThreadInfo; }
+	return static_cast<__cxa_thread_info*>(pthread_getspecific(tls_key));
+}
+
+/**
+ * This function is not a part of Itanium ABI, but most of compilers require to
+ * implement C++11 thread_local variable.
+ */
+extern "C" int __cxa_thread_atexit(void (*dtor)(void*), void *obj, void* /*dso_sym*/)
```
Comment (davidchisnall) [inline src/thread_info.cc]: This doesn't really belong here, it should be in libc.

Comment (Flast) [inline src/thread_info.cc]: IMO, libc is just a C implementation, not C++ implementation.
e.g. glibc provides `__cxa_thread_atexit_impl`, not `__cxa_thread_atexit`. libc++abi and libsupc++ provide it if glibc's `_impl` is available. Additionally, libsupc++ provides its own implementation if glibc's `_impl` is not available.

File: src/exception.cc (hunk 1 above)
Comment (Flast) [inline src/exception.cc]: OK, I move back those functions to `exception.cc`.

**Comment by cbergstrom** [comment]
@davidchisnall should this old pull request get nuked or what's the status of it?

**Comment by davidchisnall** [comment]
There's currently an approach to this in FreeBSD libc under review, I think that this is probably best to drop.

**Pull request closed without merging**